File: features/xai-oauth/src/xai_oauth/store.py
from __future__ import annotations

import fcntl
import json
import os
import tempfile
import time
import urllib.request
from collections.abc import Callable, Iterator
from contextlib import contextmanager
from pathlib import Path
from typing import Any

from xai_oauth.session import (
    AuthError,
    LoggedOut,
    OAuthSession,
    Session,
    format_expires_at,
    parse_expires_at,
)

STORE_VERSION = 1


def default_auth_path() -> Path:
    return Path.home() / ".rhc" / "auth.json"


class TokenStore:
    """Sole writer of ~/.rhc/auth.json."""

    def __init__(
        self,
        path: Path | None = None,
        *,
        opener: Callable[..., Any] | None = None,
        sleep: Callable[[float], None] | None = None,
    ) -> None:
        self.path = path if path is not None else default_auth_path()
        self._opener = opener if opener is not None else urllib.request.urlopen
        self._sleep = sleep if sleep is not None else time.sleep

    @property
    def lock_path(self) -> Path:
        return self.path.with_name(self.path.name + ".lock")

    def load(self) -> Session:
        try:
            raw = self.path.read_text(encoding="utf-8")
        except FileNotFoundError:
            return LoggedOut()
        except OSError as e:
            raise AuthError(f"cannot read {self.path}") from e
        if not raw.strip():
            # Empty file is a crash leftover (truncate before replace), not garbage.
            return LoggedOut()
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as e:
            raise AuthError(f"corrupt auth store: {self.path}") from e
        session = _session_from_wire(data, self.path)
        if isinstance(session, OAuthSession):
            return session.bind(self, self._opener, self._sleep)
        return session

    def save(self, session: OAuthSession) -> None:
        payload = json.dumps(_to_wire(session), indent=2) + "\n"
        data = payload.encode("utf-8")
        parent = self.path.parent
        try:
            parent.mkdir(parents=True, exist_ok=True)
            os.chmod(parent, 0o700)
        except OSError as e:
            raise AuthError(f"cannot create {parent}") from e
        fd = -1
        tmp = ""
        try:
            fd, tmp = tempfile.mkstemp(
                prefix=".auth.json.",
                suffix=".tmp",
                dir=parent,
            )
            os.write(fd, data)
            os.fsync(fd)
            os.close(fd)
            fd = -1
            os.chmod(tmp, 0o600)
            os.replace(tmp, self.path)
            tmp = ""
            dir_fd = os.open(parent, os.O_DIRECTORY)
            try:
                os.fsync(dir_fd)
            finally:
                os.close(dir_fd)
        except OSError as e:
            raise AuthError(f"cannot write {self.path}") from e
        finally:
            if fd >= 0:
                os.close(fd)
            if tmp:
                try:
                    os.unlink(tmp)
                except FileNotFoundError:
                    pass

    def clear(self) -> None:
        try:
            self.path.unlink()
        except FileNotFoundError:
            return
        except OSError as e:
            raise AuthError(f"cannot clear {self.path}") from e

    @contextmanager
    def lock(self) -> Iterator[None]:
        parent = self.path.parent
        try:
            parent.mkdir(parents=True, exist_ok=True)
            os.chmod(parent, 0o700)
        except OSError as e:
            raise AuthError(f"cannot create {parent}") from e
        try:
            fd = os.open(self.lock_path, os.O_CREAT | os.O_RDWR, 0o600)
        except OSError as e:
            raise AuthError(f"cannot open {self.lock_path}") from e
        try:
            fcntl.flock(fd, fcntl.LOCK_EX)
            yield
        finally:
            fcntl.flock(fd, fcntl.LOCK_UN)
            os.close(fd)


def _to_wire(session: OAuthSession) -> dict[str, Any]:
    return {
        "version": STORE_VERSION,
        "credential": {
            "kind": "oauth",
            "access_token": session.access_token,
            "refresh_token": session.refresh_token,
            "expires_at": format_expires_at(session.expires_at),
            "email": session.email,
            "subject": session.subject,
            "issuer": session.issuer,
            "client_id": session.client_id,
        },
    }


def _session_from_wire(data: object, path: Path) -> Session:
    if not isinstance(data, dict):
        raise AuthError(f"corrupt auth store: {path}")
    if data.get("version") != STORE_VERSION:
        raise AuthError(f"unsupported auth.json version: {data.get('version')!r}")
    cred = data.get("credential")
    if not isinstance(cred, dict):
        raise AuthError(f"auth.json missing credential: {path}")
    if cred.get("kind") != "oauth":
        raise AuthError(f"unsupported credential kind: {cred.get('kind')!r}")
    access = _require_str(cred, "access_token", path)
    subject = _require_str(cred, "subject", path)
    issuer = _require_str(cred, "issuer", path)
    client_id = _require_str(cred, "client_id", path)
    expires_raw = cred.get("expires_at")
    if not isinstance(expires_raw, str) or not expires_raw:
        raise AuthError(f"auth.json missing expires_at: {path}")
    refresh = cred.get("refresh_token")
    if refresh is not None and not isinstance(refresh, str):
        raise AuthError(f"auth.json has invalid refresh_token: {path}")
    if refresh == "":
        refresh = None
    email = cred.get("email")
    if email is not None and not isinstance(email, str):
        raise AuthError(f"auth.json has invalid email: {path}")
    if email == "":
        email = None
    return OAuthSession(
        access_token=access,
        refresh_token=refresh,
        expires_at=parse_expires_at(expires_raw),
        subject=subject,
        email=email,
        issuer=issuer,
        client_id=client_id,
    )


def _require_str(cred: dict[str, Any], key: str, path: Path) -> str:
    value = cred.get(key)
    if not isinstance(value, str) or not value:
        raise AuthError(f"auth.json missing {key}: {path}")
    return value

File: crates/rhc/src/host/extensions/auth/store.rs
use std::fs::{self, File, OpenOptions};
use std::io::Write;
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Utc};
use fs4::fs_std::FileExt;
use serde::Serialize;
use serde_json::Value;

use super::session::OAuthSession;
use super::xai_client::{CLIENT_ID, ISSUER};
use crate::host::ports::auth::AuthError;

const STORE_VERSION: u32 = 1;

#[derive(Serialize)]
struct StoreFile {
    version: u32,
    credential: CredentialWire,
}

#[derive(Serialize)]
struct CredentialWire {
    kind: String,
    access_token: String,
    refresh_token: Option<String>,
    expires_at: String,
    email: Option<String>,
    subject: String,
    issuer: String,
    client_id: String,
}

pub(crate) struct TokenStore {
    path: PathBuf,
}

pub(crate) struct AuthFileLock {
    _file: File,
}

impl TokenStore {
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self { path: path.into() }
    }

    pub fn in_dir(rhc_home: &Path) -> Self {
        Self::new(rhc_home.join("auth.json"))
    }

    pub fn lock_path(&self) -> PathBuf {
        self.path.with_extension("json.lock")
    }

    pub fn load(&self) -> Result<Option<OAuthSession>, AuthError> {
        let raw = match fs::read_to_string(&self.path) {
            Ok(s) => s,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(None),
            Err(e) => return Err(e.into()),
        };
        if raw.trim().is_empty() {
            return Ok(None);
        }
        let data: Value = serde_json::from_str(&raw)
            .map_err(|e| AuthError::msg(format!("corrupt auth store: {e}")))?;
        session_from_wire(&data, &self.path).map(Some)
    }

    pub fn save(&self, session: &OAuthSession) -> Result<(), AuthError> {
        let parent = self
            .path
            .parent()
            .ok_or_else(|| AuthError::msg("auth path has no parent"))?;
        fs::create_dir_all(parent)?;
        fs::set_permissions(parent, fs::Permissions::from_mode(0o700))?;
        let payload = serde_json::to_vec_pretty(&to_wire(session))
            .map_err(|e| AuthError::msg(format!("serialize auth store: {e}")))?;
        let mut tmp = tempfile_in(parent)?;
        tmp.file.write_all(&payload)?;
        tmp.file.sync_all()?;
        fs::set_permissions(&tmp.path, fs::Permissions::from_mode(0o600))?;
        fs::rename(&tmp.path, &self.path)?;
        tmp.keep = true;
        if let Ok(dir) = File::open(parent) {
            let _ = dir.sync_all();
        }
        Ok(())
    }

    pub fn clear(&self) -> Result<(), AuthError> {
        match fs::remove_file(&self.path) {
            Ok(()) => Ok(()),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(e) => Err(e.into()),
        }
    }

    pub fn lock(&self) -> Result<AuthFileLock, AuthError> {
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent)?;
            fs::set_permissions(parent, fs::Permissions::from_mode(0o700))?;
        }
        let file = OpenOptions::new()
            .create(true)
            .read(true)
            .write(true)
            .truncate(false)
            .open(self.lock_path())?;
        file.lock_exclusive()?;
        Ok(AuthFileLock { _file: file })
    }
}

impl Drop for AuthFileLock {
    fn drop(&mut self) {
        let _ = FileExt::unlock(&self._file);
    }
}

struct Tmp {
    path: PathBuf,
    file: File,
    keep: bool,
}

impl Drop for Tmp {
    fn drop(&mut self) {
        if !self.keep {
            let _ = fs::remove_file(&self.path);
        }
    }
}

fn tempfile_in(parent: &Path) -> Result<Tmp, AuthError> {
    let nonce = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    let path = parent.join(format!(".auth.json.{nonce}.tmp"));
    let file = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&path)?;
    Ok(Tmp {
        path,
        file,
        keep: false,
    })
}

fn to_wire(session: &OAuthSession) -> StoreFile {
    StoreFile {
        version: STORE_VERSION,
        credential: CredentialWire {
            kind: "oauth".into(),
            access_token: session.access_token.clone(),
            refresh_token: session.refresh_token.clone(),
            expires_at: session
                .expires_at
                .to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
            email: session.email.clone(),
            subject: session.subject.clone(),
            issuer: session.issuer.clone(),
            client_id: session.client_id.clone(),
        },
    }
}

fn session_from_wire(data: &Value, path: &Path) -> Result<OAuthSession, AuthError> {
    let obj = data
        .as_object()
        .ok_or_else(|| AuthError::msg(format!("corrupt auth store: {}", path.display())))?;
    let version = obj.get("version").and_then(Value::as_u64);
    if version != Some(STORE_VERSION as u64) {
        return Err(AuthError::msg(format!(
            "unsupported auth.json version: {:?}",
            obj.get("version")
        )));
    }
    let cred = obj
        .get("credential")
        .and_then(Value::as_object)
        .ok_or_else(|| {
            AuthError::msg(format!("auth.json missing credential: {}", path.display()))
        })?;
    if cred.get("kind").and_then(Value::as_str) != Some("oauth") {
        return Err(AuthError::msg(format!(
            "unsupported credential kind: {:?}",
            cred.get("kind")
        )));
    }
    let access = require_str(cred, "access_token", path)?;
    let subject = require_str(cred, "subject", path)?;
    let issuer = require_str(cred, "issuer", path).unwrap_or_else(|_| ISSUER.to_string());
    let client_id = require_str(cred, "client_id", path).unwrap_or_else(|_| CLIENT_ID.to_string());
    let expires_raw = require_str(cred, "expires_at", path)?;
    let expires_at = DateTime::parse_from_rfc3339(&expires_raw)
        .map_err(|e| AuthError::msg(format!("invalid expires_at: {e}")))?
        .with_timezone(&Utc);
    let refresh = opt_str(cred.get("refresh_token"));
    let email = opt_str(cred.get("email"));
    Ok(OAuthSession {
        access_token: access,
        refresh_token: refresh,
        expires_at,
        subject,
        email,
        issuer,
        client_id,
    })
}

fn require_str(
    cred: &serde_json::Map<String, Value>,
    key: &str,
    path: &Path,
) -> Result<String, AuthError> {
    match cred.get(key).and_then(Value::as_str) {
        Some(s) if !s.is_empty() => Ok(s.to_string()),
        _ => Err(AuthError::msg(format!(
            "auth.json missing {key}: {}",
            path.display()
        ))),
    }
}

fn opt_str(value: Option<&Value>) -> Option<String> {
    value
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

#[cfg(test)]
#[path = "store_tests.rs"]
mod tests;
